ibis/backends/datafusion/__init__.py

```diff
@@ -0,0 +1,50 @@
+import pandas as pd
+import toolz
+from pkg_resources import parse_version
+
+import ibis.expr.operations as ops
+import ibis.expr.schema as sch
+from ibis.backends.base import BaseBackend
+from ibis.backends.base.file import FileClient
+from ibis.expr.scope import Scope
+from ibis.expr.typing import TimeContext
+
+from .client import DataFusionClient, DataFusionDatabase, DataFusionTable
+from .compiler import DataFusionExprTranslator
+
+
+class Backend(BaseBackend):
+    name = 'datafusion'
+    builder = None
+    client_class = DataFusionClient
+    database_class = DataFusionDatabase
+    table_class = DataFusionTable
+    translator_class = DataFusionExprTranslator
+
+    def version(self):
+        return '1.0.0dev'
+
+    def connect(self, config):
+        """
+        Create a DataFusionClient for use with Ibis
+
+        Parameters
+        ----------
+        config : DataFusionContext or dict
+
+        Returns
+        -------
+        DataFusionClient
+        """
+        return self.client_class(backend=self, config=config)
+
+    def current_database(self):
+        raise NotImplementedError("")
```

Comment:
Should we go ahead and make `BaseBackend` an abstract base class so you don't have to have pointless boilerplate here?

Reply:
There was some inconsistent behavior with this until #2911. The options I considered were returning `None` by default, and let backends supporting different databases overwrite it. And requiring this, expecting all backends to support this, even if only one database is supported (this can return `base` or `main`). SQLAlchemy is using the latter, and I think it makes things simpler (including tests and type annotations). This way we can garantee code like `assert client.current_database in client.list_databases()` to work for all backends, which I think it's nice for users and keeps things simpler than having to handle `None` and special cases.

Happy to discuss further (in a new issue), but unless we change that, you can return `main` or whatever here, and `['main']` in `list_databases()`.

Reply:
Updated to use the new backend abstract class which is way nicer to use!